src/hashupdater.cpp

```diff
@@ -0,0 +1,100 @@
+#include "hashupdater.hpp"
+#include "util.hpp"
+#include "world.hpp"
+
+PasswordHashUpdater::PasswordHashUpdater(Config& config, const std::unordered_map<HashFunc, std::shared_ptr<Hasher>>& passwordHashers)
+    : _config(config)
+    , _passwordHashers(passwordHashers)
+    , _database(nullptr)
+    , _terminating(false)
+{
+    this->_updateThread = std::thread([this] { this->updateThreadProc(); });
+
+    auto dbType = util::lowercase(std::string(config["DBType"]));
+
+    Database::Engine engine = Database::MySQL;
+    if (!dbType.compare("sqlite"))
+    {
+        engine = Database::SQLite;
+    }
+    else if (!dbType.compare("sqlserver"))
+    {
+        engine = Database::SqlServer;
+    }
+    else if (!dbType.compare("mysql"))
+    {
+        engine = Database::MySQL;
+    }
+
+    auto dbHost = std::string(config["DBHost"]);
+    auto dbUser = std::string(config["DBUser"]);
+    auto dbPass = std::string(config["DBPass"]);
+    auto dbName = std::string(config["DBName"]);
+    auto dbPort = int(config["DBPort"]);
+
+    this->_database.reset(new Database(engine, dbHost, dbPort, dbUser, dbPass, dbName));
+}
+
+PasswordHashUpdater::~PasswordHashUpdater()
+{
+    this->_terminating = true;
+
+    this->_updateSignal.notify_all();
+    this->_updateThread.join();
+}
+
+void PasswordHashUpdater::QueueUpdatePassword(const std::string& username, util::secure_string&& password, HashFunc hashFunc)
+{
+    UpdateState state { username, std::move(password), hashFunc };
+
+    std::lock_guard<std::mutex> queueGuard(this->_updateQueueLock);
+    this->_updateQueue.push(std::move(state));
+
+    this->_updateSignal.notify_one();
+}
+
+void PasswordHashUpdater::updateThreadProc()
+{
+    while (!this->_terminating)
+    {
+        std::unique_lock<std::mutex> lock(this->_updateMutex);
+        this->_updateSignal.wait(lock);
+
+        if (this->_terminating)
+        {
+            break;
+        }
```

Comment:
What's the reason for the break here? My guess is that it was added for the scenario where the `_updateMutex` lock is released as a result of termination to prevent additional actions from taking place once a user has initiated termination?

Reply:
That's exactly correct. Termination requests signal the event (via destructor), and we want to quit as quickly as possible so it skips any pending password updates.